pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net"
+	"net/netip"
+	"slices"
+
+	"github.com/cilium/hive/cell"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+
+	"github.com/cilium/cilium/pkg/clustermesh/types"
+	"github.com/cilium/cilium/pkg/counter"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/fqdn/messagehandler"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/time"
+
+	pb "github.com/cilium/cilium/api/v1/standalone-dns-proxy"
+)
+
+// FQDNDataServer is the server for the standalone DNS proxy grpc server
+// It is responsible for handling the FQDN mapping requests from the SDP
+// and sending the DNS Policy updates to the SDP.
+type FQDNDataServer struct {
+	pb.UnimplementedFQDNDataServer
+
+	// port is the port on which the standalone DNS proxy grpc server will run
+	port int
+
+	// grpcServer is the grpc server for the standalone DNS proxy
+	grpcServer *grpc.Server
+
+	endpointManager endpointmanager.EndpointManager
+
+	// updateOnDNSMsg is a function to update the DNS message in the cilium agent on receiving the FQDN mapping
+	updateOnDNSMsg messagehandler.DNSMessageHandler
+
+	// identityToIPMutex is a mutex to protect the current state of the identity to Ip mapping
+	identityToIPMutex lock.Mutex
+
+	// currentIdentityToIP is a map of the identity to list of IPs
+	currentIdentityToIP map[identity.NumericIdentity][]netip.Prefix
+
+	// prefixLengths tracks the unique set of prefix lengths for IPv4 and
+	// IPv6 addresses in order to optimize longest prefix match lookups.
+	prefixLengths *counter.PrefixLengthCounter
+
+	// log is the logger for the FQDNDataServer
+	log *slog.Logger
+
+	// listener is used to create a net.Listener when starting the grpc server
+	listener Listener
+}
+
+var (
+	kaep = keepalive.EnforcementPolicy{
+		PermitWithoutStream: true, // Allow pings even when there are no active streams
+	}
+	kasp = keepalive.ServerParameters{
+		Time:    5 * time.Second, // Ping the client if it is idle for 5 seconds to ensure the connection is still active
+		Timeout: 1 * time.Second, // Wait 1 second for the ping ack before assuming the connection is dead
+	}
+)
+
+// Listener is an interface that abstracts the creation of a net.Listener.
+type Listener interface {
```

Comment:
nit: `Listener` can (and should be) private - this is an implementation detail. While you're at it - maybe `Listener` isn't such a great name after all; given it _returns_ a `net.Listener`. Could go with `listenConfig` like the stdlib.